proposals/p2360.md

```diff
@@ -0,0 +1,553 @@
+  d.((Base as Iface).Method)();
+}
+```
+
+### Tuple and struct literals
+
+A tuple literal such as `(1, 2, 3)` produces a tuple value. Its type cannot be
+written directly as a literal: the literal `(i32, i32, i32)` also produces a
+tuple value, not a type value. However, the type of `(1, 2, 3)` is easy to
+express: it is the result of implicitly converting `(i32, i32, i32)` to a type,
+so for example `(i32, i32, i32) as type` evaluates to the type of `(1, 2, 3)`.
+
+There is no conversion from a tuple type to a tuple value, because tuple types
+are values of type `type`, and the type `type` does not implicitly convert to a
+tuple type. For metaprogramming, it is likely that we will benefit from having a
+mechanism to convert a tuple type into a tuple of types, but this is likely to
+be possible using a variadic:
+
+```
+// Takes a tuple of values. Returns a tuple containing the types of those values.
+fn TupleTypeToTupleOfTypes[Types:! type,...](x: (Types,...)) -> auto {
+  return (Types,...);
+}
+```
+
+### Name lookup into values with constrained types
+
+Given a generic function with a generic type parameter, uses of that parameter
+will be implicitly converted to type `type`. For example:
+
+```
+fn F[T:! Printable](x: T) { x.Print(); }
+```
+
+... is equivalent to ...
+
+```
+fn F[T:! Printable](x: T as type) { x.Print(); }
+```
+
+As a result, we can no longer describe lookup for `x` as looking into the
+type-of-type, that is, the type of the expression after `x:`, because after
+implicit conversion, that expression is of type `type`. Instead, lookup will
+look into the type of `x`, which symbolically evaluates to the value
+corresponding to `T` in type `type`, and will look at that value to determine
+where else to look. Because that value is symbolically the name of a generic
+type parameter, we will look in the type of `T`, which is `Printable`.
+
+## Rationale
+
+-   [Language tools and ecosystem](/docs/project/goals.md#language-tools-and-ecosystem)
+    -   Making the behavior of tuples and empty structs more like other types is
+        easier for tooling to handle.
+-   [Software and language evolution](/docs/project/goals.md#software-and-language-evolution)
+    -   The transition between templates and generics is made smoother by
+        changing compound member access to behave consistently regardless of
+        whether the left-hand operand is a type.
+-   [Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)
+    -   Bottoming out at the same type, `type`, for all types is likely to be
+        easier to understand. However, having `(i32, i32)` not be the type of a
+        pair of integers, but instead be _implicitly convertible to_ the type of
+        a pair of integers, is likely to be harder to understand. It's unclear
+        which of these will be more dominant, but the new behavior is more
+        consistent.
+
+## Alternatives considered
```

Comment:
I think we should also capture the alternative of using two syntaxes such as:

```
var x: TupleType(i32, i32) = (1, 2);

// Or, with dedicated syntax:
var y: (|i32, i32|) = (1, 2);
```

That at least came up with some decent about of discussion with @camio and others I think, and its an important alternative to call out here. Others will wonder about it and we should make it clear that this wasn't just missed.

Reply:
I added this at the bottom.